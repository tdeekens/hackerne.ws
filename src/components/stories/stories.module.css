.stories {
   display: flex;
   flex-direction: column;
   align-items: center;
   justify-content: center;
   margin: 12px 0;
}

.stories > * + * {
   border-bottom: 1px solid #f2f3f5
}

.loadingStateIndicator {
   padding: 12px 0;
}
