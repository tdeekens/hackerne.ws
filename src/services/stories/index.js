export { default as createStoriesService } from './stories';
