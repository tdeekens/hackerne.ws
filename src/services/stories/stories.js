const defaultConfig = {
  uri: 'https://hacker-news.firebaseio.com/v0/item',
  format: 'json',
};

/**
 * Service to fetch one or multiple stories from the hackernews API.
 *
 * It currently supports fetching one `fetchOne` story by id. The default
 * configuration can be overwritten by passing in a custom config.
 */
export default function create(config = defaultConfig) {
  const fetchOne = id =>
    fetch(`${config.uri}/${id}.${config.format}`).then(response =>
      response.json()
    );

  return {
    fetchOne,
  };
}
