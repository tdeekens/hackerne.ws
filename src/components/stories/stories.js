import React from 'react';
import Story from '../story';
import InfiniteScroll from '../infinite-scroll';
import { createTopStoriesService } from '../../services/top-stories';
import styles from './stories.module.css';

export class Stories extends React.Component {
  static displayName = 'Stories';

  service = createTopStoriesService();

  state = {
    stories: [],
    isLoading: false,
    hasMore: true,
    error: null,
  };

  setStory = story => {
    this.setState(prevState => ({
      stories: [...prevState.stories, story],
    }));
  };

  setError = error => {
    this.setState({ error });
  };

  setIsLoading = isLoading => {
    this.setState({ isLoading });
  };

  setHasMore = hasMore => {
    this.setState({ hasMore });
  };

  loadingSentinalRef = React.createRef();

  handleFetchMore = () => {
    this.setIsLoading(true);

    this.service.fetchMore();
  };

  componentDidMount() {
    this.service
      .subscribe(
        ({ payload: story, hasMore }) => {
          this.setStory(story);
          this.setHasMore(hasMore);
          this.setIsLoading(false);
        },
        error => {
          this.setError(error);
        }
      )
      .then(() => {
        this.handleFetchMore();
      });
  }

  render() {
    return (
      <div className={styles.stories}>
        <InfiniteScroll
          isLoading={this.state.isLoading}
          onFetchMore={this.handleFetchMore}
          sentinelRef={this.loadingSentinalRef}
        >
          {this.state.stories.map(story => (
            <Story
              key={story.id}
              id={story.id}
              author={story.by}
              title={story.title}
              url={story.url}
              score={story.score}
              publishedAt={story.time}
            />
          ))}
          <div
            ref={this.loadingSentinalRef}
            className={styles.loadingStateIndicator}
          >
            {this.state.isLoading && <span>Loading...</span>}
            {!this.state.hasMore && <span>No more stories...</span>}
            {this.state.error && <span>Sorry, something went wrong...</span>}
          </div>
        </InfiniteScroll>
      </div>
    );
  }
}

export default Stories;
