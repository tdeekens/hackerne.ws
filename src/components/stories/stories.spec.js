import React from 'react';
import { shallow } from 'enzyme';
import { Stories } from './stories';
import InfiniteScroll from '../infinite-scroll';
import Story from '../story';

const createStory = custom => ({
  id: 1,
  by: 'John Doe',
  title: 'Test story title',
  url: 'http://test-url.com',
  score: 12,
  time: 1541410855801,
  ...custom,
});

beforeAll(() => {
  global.fetch = jest.fn(() => Promise.resolve());
});

describe('rendering', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Stories />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('stories', () => {
    beforeEach(() => {
      wrapper.instance().setStory(createStory());
      wrapper.instance().setStory(createStory({ id: 2 }));
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render each <Story>', () => {
      expect(wrapper).toContainMatchingElements(2, Story);
    });
  });

  describe('fetching states', () => {
    describe('when loading', () => {
      beforeEach(() => {
        wrapper.instance().setIsLoading(true);
      });

      it('should give a loading state indication', () => {
        expect(wrapper.find('span')).toHaveText('Loading...');
      });
    });

    describe('when not having more', () => {
      beforeEach(() => {
        wrapper.instance().setHasMore(false);
      });

      it('should give a indication of not having more', () => {
        expect(wrapper.find('span')).toIncludeText('No more stories');
      });
    });

    describe('when error occured', () => {
      beforeEach(() => {
        wrapper.instance().setError(new Error('test error'));
      });

      it('should give a indication of not having more', () => {
        expect(wrapper.find('span')).toIncludeText(
          'Sorry, something went wrong'
        );
      });
    });
  });
});

describe('state handlers', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Stories />);
  });

  describe('setStory', () => {
    beforeEach(() => {
      wrapper.instance().setStory(createStory());
      wrapper.instance().setStory(createStory({ id: 2 }));
    });

    it('should add each added story', () => {
      expect(wrapper.state('stories')).toHaveLength(2);
    });
  });

  describe('setError', () => {
    const error = new Error('test error');
    beforeEach(() => {
      wrapper.instance().setError(error);
    });

    it('should set the error', () => {
      expect(wrapper).toHaveState('error', error);
    });
  });

  describe('setIsLoading', () => {
    beforeEach(() => {
      wrapper.instance().setIsLoading(true);
    });

    it('should set the loading state', () => {
      expect(wrapper).toHaveState('isLoading', true);
    });
  });

  describe('setHasMore', () => {
    beforeEach(() => {
      wrapper.instance().setHasMore(false);
    });

    it('should set the state', () => {
      expect(wrapper).toHaveState('hasMore', false);
    });
  });
});

describe('lifecycles', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Stories />);

    jest.spyOn(wrapper.instance().service, 'subscribe');

    wrapper.instance().componentDidMount();
  });

  it('should subscribe for changes on the service', () => {
    expect(wrapper.instance().service.subscribe).toHaveBeenCalled();

    expect(wrapper.instance().service.subscribe).toHaveBeenCalledWith(
      expect.any(Function),
      expect.any(Function)
    );
  });
});
