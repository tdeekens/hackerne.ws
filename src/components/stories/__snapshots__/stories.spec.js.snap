// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should match snapshot 1`] = `
<div>
  <InfiniteScroll
    isLoading={false}
    onFetchMore={[Function]}
    sentinelRef={
      Object {
        "current": null,
      }
    }
  >
    <div />
  </InfiniteScroll>
</div>
`;

exports[`rendering stories should match snapshot 1`] = `
<div>
  <InfiniteScroll
    isLoading={false}
    onFetchMore={[Function]}
    sentinelRef={
      Object {
        "current": null,
      }
    }
  >
    <Story
      author="John Doe"
      id={1}
      key="1"
      publishedAt={1541410855801}
      score={12}
      title="Test story title"
      url="http://test-url.com"
    />
    <Story
      author="John Doe"
      id={2}
      key="2"
      publishedAt={1541410855801}
      score={12}
      title="Test story title"
      url="http://test-url.com"
    />
    <div />
  </InfiniteScroll>
</div>
`;
